deep_speech_2/decoders/swig/ctc_decoders.cpp

```diff
@@ -0,0 +1,332 @@
+#include "ctc_decoders.h"
+
+#include <algorithm>
+#include <cmath>
+#include <iostream>
+#include <limits>
+#include <map>
+#include <utility>
+
+#include "ThreadPool.h"
+#include "fst/fstlib.h"
+
+#include "decoder_utils.h"
+#include "path_trie.h"
+
+std::string ctc_greedy_decoder(
+    const std::vector<std::vector<double>> &probs_seq,
+    const std::vector<std::string> &vocabulary) {
+  // dimension check
+  int num_time_steps = probs_seq.size();
+  for (int i = 0; i < num_time_steps; i++) {
+    if (probs_seq[i].size() != vocabulary.size() + 1) {
+      std::cout << "The shape of probs_seq does not match"
+                << " with the shape of the vocabulary!" << std::endl;
+      exit(1);
+    }
+  }
+
+  int blank_id = vocabulary.size();
+
+  std::vector<int> max_idx_vec;
+  double max_prob = 0.0;
+  int max_idx = 0;
+  for (int i = 0; i < num_time_steps; i++) {
+    for (int j = 0; j < probs_seq[i].size(); j++) {
+      if (max_prob < probs_seq[i][j]) {
+        max_idx = j;
+        max_prob = probs_seq[i][j];
+      }
+    }
+    max_idx_vec.push_back(max_idx);
+    max_prob = 0.0;
+    max_idx = 0;
+  }
+
+  std::vector<int> idx_vec;
+  for (int i = 0; i < max_idx_vec.size(); i++) {
+    if ((i == 0) || ((i > 0) && max_idx_vec[i] != max_idx_vec[i - 1])) {
+      idx_vec.push_back(max_idx_vec[i]);
+    }
+  }
+
+  std::string best_path_result;
+  for (int i = 0; i < idx_vec.size(); i++) {
+    if (idx_vec[i] != blank_id) {
+      best_path_result += vocabulary[idx_vec[i]];
+    }
+  }
+  return best_path_result;
+}
+
+std::vector<std::pair<double, std::string>> ctc_beam_search_decoder(
+    const std::vector<std::vector<double>> &probs_seq,
+    int beam_size,
+    std::vector<std::string> vocabulary,
```

Comment:
const &

Reply:
Did not use ```const &``` because there would be an error when using ```std::find``` for this variable. And hasn't fixed this problem yet.